### 16S rRNA gene amplicon analysis using qiime2 on Quest

By [**Matt Selensky**](https://mselensky.github.io/)

*31 May 2023*

This document demonstrates an example workflow to process 16S rRNA gene data from multiplexed paired-end FASTQ files using Northwestern's high-performance computing cluster, [Quest](https://services.northwestern.edu/TDClient/30/Portal/KB/ArticleDet?ID=1542). Before you proceed, make sure you are familiar with the basics of using Quest have access to the Osburn Lab group Allocation I ([`p31618`](https://app.smartsheet.com/b/form/797775d810274db5889b5199c4260328))! 

This workflow uses version 2022.2 of the [qiime2](https://qiime2.org/) software to demultiplex, trim, quality-control, identify ASVs, assign taxonomy, build a phylogenetic tree, export an ASV table, as well as visualize rarefaction plots, taxa barplots, and statistics from many of the intermediate steps. These steps are executed by three separate bash scripts submitted to Quest using the `p31618` allocation. The fourth script exports the Qiime2-formatted data (.qza extensions) into ASV-level feature tables (.biom and .txt), FASTA-formatted representative sequences, and demultiplexed FASTQ files for downstream purposes. You will have to modify each script with the path to the input fastq files (marked in the scripts with `# change to your data directory`) and your email address to be notified when each job completes (marked with `# change to your email`). Be sure to modify each function's parameters as is most appropriate for your dataset. For debugging purposes, refer to the `.out` files generated by each job submission.

To set up the workflow environment, first clone this repository and navigate to the `16S_pipeline_qiime2` folder:

```
git clone https://github.com/NU-OsburnLab/Example_Scripts.git
cd Example_Scripts/16S_pipeline_qiime2
```

Next, modify each script as necessary (`nano` or `vi` are available as text editors on Quest). Side note - if you are a member of `p31618`, you should have permission to activate the `qiime2-2022.2` conda environment without modification.

To run the workflow, submit the following scripts in order:

##### `q2-import-demux-cutadapt.sh`

This script performs:

1. Importing paired-end, multiplexed reads into qiime2
2. Demultiplexing imported reads with the [cutadapt plugin](https://docs.qiime2.org/2022.2/plugins/available/cutadapt/)
3. Trimming primer sequences (from 515FY/926R primers, [Parada et al. 2015](https://sfamjournals.onlinelibrary.wiley.com/doi/abs/10.1111/1462-2920.13023)) from demultiplexed reads with the cutadapt plugin

This script assumes you have a [qiime2-formatted metadata file](https://docs.qiime2.org/2022.2/tutorials/metadata/) saved in your working directory as `metadata.tsv`, with barcode sequences listed in a column named `barcode`. 

Once the appropriate modifications are made to this script, submit a job to Quest with the command `sbatch q2-import-demux-cutadapt.sh`. 

After the job completes, download the output file `qiime2-out/demux-trimmed-no-untrimmed.qzv` to your computer and visualize it on [view.qiime2.org](https://view.qiime2.org). [Globus](https://kb.northwestern.edu/page.php?id=70521) is a useful tool to transfer this file to your computer from Quest. Note where the quality scores drop off in each read from the interactive quality plot before moving on to the next script.

##### `q2-dada2.sh`

This script performs:

4. Denoising and ASV assigments via the [DADA2 plugin](https://docs.qiime2.org/2022.2/plugins/available/dada2/)

Modify the `--p-trim` and `--p-trunc-len` parameters based on the results from visualizing `demux-trimmed-no-untrimmed.qzv`. Refer to the DADA2 plugin's [documentation](https://docs.qiime2.org/2022.2/plugins/available/dada2/denoise-paired/) for more information.

Once the appropriate modifications are made to this script, submit a job to Quest with the command `sbatch q2-dada2.sh`.

##### `q2-taxonomy-phylogenetic-tree.sh`

This script performs:

5. Taxonomy assignment with custom pre-trained Silva 138 classifier via [classify-sklearn](https://docs.qiime2.org/2020.2/plugins/available/feature-classifier/classify-sklearn/)
6. Creating rooted and unrooted phylogenetic trees via [fasttree and mafft alignment](https://docs.qiime2.org/2022.2/plugins/available/phylogeny/align-to-tree-mafft-fasttree/)
7. Visualizing rarefaction plots
8. Collapsing ASV tables for easier import into Python/R
9. Visualizing taxonomic composition barplot

Once the appropriate modifications are made to this script, submit a job to Quest with the command `sbatch q2-taxonomy-phylogenetic-tree.sh` once the previous DADA2 job finishes. If you don't want to wait to submit this job, you can tell SLURM to run it once the DADA2 job successfully completes via: `sbatch --dependency:afterok=XXXXXX q2-taxonomy-phylogenetic-tree.sh`, where `XXXXXX` is the numeric ID of the DADA2 job. If you didn't catch it when you first submitted the job, note the ID for the one named "q2-dada2" after running `squeue -u $USER` and use that.

##### `q2-export_data.sh`

This script exports useful files from Qiime2 (ending with *.qza*) to those with more common extensions for downstream purposes:

10. Exports biom-/txt-formatted feature tables, taxonomies, and fasta-formatted representative sequences to the `feature_tables` output subfolder. 
11. Exports demultiplexed forward and reverse FASTQ files to the `fastqs` output subfolder.